# 综合题(选做)

任务要求：白葡萄酒质量共分为7个等级，对白葡萄酒质量进行预测。
1. 使用精度作为评价指标。
2. 使用一元线性回归、多元线性回归、对数线性回归等线性回归模型对葡萄酒质量进行预测，计算其十折交叉验证的精度。
3. 使用对数几率回归、线性判别分析对葡萄酒质量进行预测，计算其十折交叉验证的精度。
4. 对比不同模型的预测能力。

## 读取数据

In [1]:
import pandas as pd
data = pd.read_csv('data/wine_quality/winequality-white.csv', delimiter=";")

In [2]:
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [3]:
# YOUR CODE HERE
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict
import numpy as np


In [4]:
x = data.iloc[:,:-1]
y = data.iloc[:,-1]

In [5]:
# 一元线性回归
acc_list = []
pre_list = []
rec_list = []
f1_list = []
lr = LinearRegression()
for i in range(0, 11):
    prediction = cross_val_predict(lr, x.iloc[:, i].values.reshape(-1,1), y, cv = 10)
    prediction = np.round(prediction)
    acc_list.append(accuracy_score(y, prediction))
    pre_list.append(precision_score(y, prediction, average="weighted"))
    rec_list.append(recall_score(y, prediction, average="weighted"))
    f1_list.append(f1_score(y, prediction, average="weighted"))
idx = f1_list.index(max(f1_list))
print(acc_list[idx])
print(pre_list[idx])
print(rec_list[idx])
print(f1_list[idx])

0.482850142915
0.446476657533
0.482850142915
0.434697110014


/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anacon

In [6]:
# 多元线性回归
lr = LinearRegression()

prediction = cross_val_predict(lr, x, y, cv = 10)
prediction = np.round(prediction)
print(accuracy_score(y, prediction))
print(precision_score(y, prediction, average="weighted"))
print(recall_score(y, prediction, average="weighted"))
print(f1_score(y, prediction, average="weighted"))

0.513270722744
0.492855029017
0.513270722744
0.475905889689


/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1137: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


In [7]:
print(acc_list)
print(pre_list)
print(rec_list)
print(f1_list)

[0.44814209881584322, 0.4540628828093099, 0.44875459371171905, 0.44814209881584322, 0.44997958350347078, 0.44875459371171905, 0.44793793385055125, 0.44855042874642709, 0.44875459371171905, 0.44875459371171905, 0.48285014291547568]
[0.23852660741290008, 0.32566978345229669, 0.20138068537737006, 0.20122907726194553, 0.35181978027811062, 0.20138068537737006, 0.32465286410207927, 0.50627439024651177, 0.20138068537737006, 0.20138068537737006, 0.44647665753332005]
[0.44814209881584322, 0.4540628828093099, 0.44875459371171905, 0.44814209881584322, 0.44997958350347078, 0.44875459371171905, 0.44793793385055125, 0.44855042874642709, 0.44875459371171905, 0.44875459371171905, 0.48285014291547568]
[0.27821862232188749, 0.30357600989941302, 0.27800524153843253, 0.27774322097764653, 0.29904308019972553, 0.27800524153843253, 0.28545198656385495, 0.33180639503982556, 0.27800524153843253, 0.27800524153843253, 0.43469711001449945]


**从上述数据可看出，alcohol和quality影响最大**

In [8]:
# 对数线性回归
prediction = cross_val_predict(lr, x, np.log(y), cv = 10)
prediction = np.exp(prediction)
prediction = np.round(prediction)
print(accuracy_score(y, prediction))
print(precision_score(y, prediction, average="weighted"))
print(recall_score(y, prediction, average="weighted"))
print(f1_score(y, prediction, average="weighted"))

0.52082482646
0.499629910333
0.52082482646
0.483102789303


/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [9]:
# 对数几率回归
logit = LogisticRegression()
prediction = cross_val_predict(logit, x, y, cv = 10)
print(accuracy_score(y, prediction))
print(precision_score(y, prediction, average="weighted"))
print(recall_score(y, prediction, average="weighted"))
print(f1_score(y, prediction, average="weighted"))

/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 5 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)


0.522866476113
0.48904146272
0.522866476113
0.464917405992


/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [10]:
# 线性判别分析
LDA = LinearDiscriminantAnalysis()
prediction = cross_val_predict(LDA, x, y, cv = 10)
print(accuracy_score(y, prediction))
print(precision_score(y, prediction, average="weighted"))
print(recall_score(y, prediction, average="weighted"))
print(f1_score(y, prediction, average="weighted"))

0.52123315639
0.497848975685
0.52123315639
0.49748412615


/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 5 members, which is too few. The minimum number of members in any class cannot be less than n_splits=10.
  % (min_groups, self.n_splits)), Warning)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:442: UserWarning: The priors do not sum to 1. Renormalizing
  UserWarning)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:442: UserWarning: The priors do not sum to 1. Renormalizing
  UserWarning)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:442: UserWarning: The priors do not sum to 1. Renormalizing
  UserWarning)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/discriminant_analysis.py:442: UserWarning: The priors do not sum to 1. Renormalizing
  UserWarning)
/Users/yangyu/anaconda3/lib/python3.6/site-packages/sklearn/discriminant

| 模型             | 精度          | 查准率           | 查全率         | F1            |
| --------------------- | ------------------ | ------------------- | ------------------ | ------------------- |
| 一元线性回归(alcohol) | 0.4828501429154757 | 0.4828501429154757  | 0.4828501429154757 | 0.43469711001449945 |
| 多元线性回归       | 0.5132707227439771 | 0.492855029017181   | 0.5132707227439771 | 0.47590588968859865 |
| 对数线性回归       | 0.5208248264597795 | 0.49962991033268606 | 0.5208248264597795 | 0.483102789303352   |
| 对数几率回归       | 0.5228664761126991 | 0.488663599650864   | 0.5228664761126991 | 0.46472415030792047 |
| 线性判别分析       | 0.5212331563903634 | 0.4978489756853334  | 0.5212331563903634 | 0.49748412615015214 |